### PR TITLE
Handle DeQuantStub() for QAT

### PR DESCRIPTION
**Pull request by raghuramank100**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26627 Default observer and fake-quant for backends
* #26626 Moving average for statistics in observers
* #26625 Emulate weight and activation only quant with fake quant, numerics test
* #26624 Quantization aware training: Freeze batch norm support
* #26623 Per channel fake quant
* #26516 Per-channel baseline
* #26520 control of observer/fake-quant operations
* #26622 Default histogram observer
* #26519 Serialization and range reduction support for Fake Quant/Observer
* #26420 Fake quantization enhancements for QAT/PTQ support
* **#26518 Handle DeQuantStub() for QAT**

Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

### CODE REVIEW

File: torch/quantization/quantize.py [hunk 1]
```diff
@@ -229,6 +232,8 @@ def forward(self, x):
     nn.Linear: nnqd.Linear,
     nn.LSTM: nnqd.LSTM,
 }
+# List of modules for which we do not need FakeQuant for activations
+DEFAULT_QAT_SKIP_MODULE_LIST = [DeQuantStub]
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: shall we do whitelist?

Comment (raghuramank100) [reply torch/quantization/quantize.py]: This would need to be done for all the lists in quantize, so that the API is consistent. Shall we do this in a later PR?

**Comment by jerryzh168** [comment]
discussed offline about removing skip_list

File: torch/quantization/quantize.py [hunk 2]
```diff
@@ -102,8 +102,9 @@ def add_observer(module):
     if hasattr(module, 'qconfig') and module.qconfig is not None and \
        len(module._modules) == 0:
         # observer and hook will be gone after we swap the module
-        module.add_module('observer', module.qconfig.activation())
-        module.register_forward_hook(_observer_forward_hook)
+        if module.training and type(module) not in set(DEFAULT_QAT_SKIP_MODULE_LIST) or module.training is False:
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: I think this is overcomplicated. just use`if not (module.training and type(module) == DeQuantStub)`

File: torch/quantization/quantize.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/quantize.py]: we can remove this

File: torch/quantization/quantize.py (hunk 2 above)
Comment (dzhulgakov) [reply torch/quantization/quantize.py]: this feels like a bit too magical to be so deep in the code. Why can't we just pass DeQuantStub as a skip list to propagate_qconfig? it already has a skip list

**Review by dzhulgakov — CHANGES_REQUESTED** [review]

**Comment by jerryzh168** [comment]
@dzhulgakov the skip only happens when the module is in training mode

**Comment by jerryzh168** [comment]
also we plan to change it to whiltelist I think

File: torch/quantization/quantize.py (hunk 2 above)
Comment (raghuramank100) [reply torch/quantization/quantize.py]: Started with that design, with a list being passed to propagate_qconfig for the case of quantization aware training.  

Since we are going to change this to a whitelist (after 1.3), we can make this change later.  We need to land this soon, so lets settle on a preferred way soon. 

My recommendation is to have a list just so that the flow is clear to the end user and have it explicitly passed within prepare_qat calls (like the original version of the diff).

File: torch/quantization/quantize.py [hunk 3]
```diff
@@ -102,8 +102,9 @@ def add_observer(module):
     if hasattr(module, 'qconfig') and module.qconfig is not None and \
        len(module._modules) == 0:
         # observer and hook will be gone after we swap the module
-        module.add_module('observer', module.qconfig.activation())
-        module.register_forward_hook(_observer_forward_hook)
+        if module.training and type(module) != DeQuantStub or module.training is False:
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: can you change this to `if not (module.training and type(module) == DeQuantStub)`

File: torch/quantization/quantize.py [hunk 4]
```diff
@@ -11,8 +11,35 @@
...
+            self.qconfig = qconfig
+
+    def forward(self, x):
+        return x
+
+class DeQuantStub(nn.Module):
+    r"""Dequantize stub module, before calibration, this is same as identity,
+    this will be swapped as `nnq.DeQuantize` in `convert`.
+    """
+    def __init__(self, qconfig=None):
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: then we don't need qconfig in DeQuantStub?

**Review by dzhulgakov — APPROVED** [review]
Looks good (maybe check Jerry's comment)

**Comment by raghuramank100** [comment]
> Looks good (maybe check Jerry's comment)

Have addressed Jerry's comment, will land after tests pass.

File: torch/quantization/quantize.py (hunk 4 above)
Comment (raghuramank100) [reply torch/quantization/quantize.py]: Yes, removed it now

**Review by jerryzh168 — APPROVED** [review]
LG Thanks
